Remove incorrect BuiltinVariable.call_hasattr()

aorenste opened the pull request:
BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`.

The BuiltinVariable version is used as a pattern from `call_self_handler()` for `BuiltinVariable(hasattr)`. I think the other version is just used for internal `hasattr(obj, name)` so I renamed that one to `call_obj_hasattr`.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #145554
* #145553
* #145552
* __->__ #145551



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

aorenste committed:
Split VariableTracker.call_hasattr() into two distinct functions.

[ghstack-poisoned]

aorenste commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -975,7 +975,7 @@ def _handle_insert_op_in_graph(self, tx: "InstructionTranslator", args, kwargs):
                     args[0], variables.UnspecializedPythonVariable
                 ):
-                    args[0] = args[0].convert_to_constant(tx)
+                    args[0] = args[0].as_python_constant()
```
This has been around since before dynamo was folded into pytorch.  `convert_to_constant` doesn't exist anywhere else in the codebase so I assume this was meant to be `as_python_constant` (and no tests seem to trigger this behavior).

aorenste committed:
Update on "Remove incorrect BuiltinVariable.call_hasattr()"

BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`. It's also only used in a single place (also within BuiltinVariable). Inline the single use and then remove it to remove confusion about which is which.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68593080).

aorenste committed:
Update on "Remove incorrect BuiltinVariable.call_hasattr()"

BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`. It's also only used in a single place (also within BuiltinVariable). Inline the single use and then remove it to remove confusion about which is which.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593080](https://our.internmc.facebook.com/intern/diff/D68593080)

[ghstack-poisoned]

aorenste committed:
Update on "Remove incorrect BuiltinVariable.call_hasattr()"

BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`. It's also only used in a single place (also within BuiltinVariable). Inline the single use and then remove it to remove confusion about which is which.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593080](https://our.internmc.facebook.com/intern/diff/D68593080)

[ghstack-poisoned]

aorenste committed:
Update on "Remove incorrect BuiltinVariable.call_hasattr()"

BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`.

The BuiltinVariable version is used as a pattern from `call_self_handler()` for `BuiltinVariable(hasattr)`. I think the other version is just used for internal `hasattr(obj, name)` so I renamed that one to `call_obj_hasattr`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593080](https://our.internmc.facebook.com/intern/diff/D68593080)

[ghstack-poisoned]

aorenste committed:
Update on "Remove incorrect BuiltinVariable.call_hasattr()"

BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`.

The BuiltinVariable version is used as a pattern from `call_self_handler()` for `BuiltinVariable(hasattr)`. I think the other version is just used for internal `hasattr(obj, name)` so I renamed that one to `call_obj_hasattr`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593080](https://our.internmc.facebook.com/intern/diff/D68593080)

[ghstack-poisoned]

aorenste committed:
Update on "Remove incorrect BuiltinVariable.call_hasattr()"

BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`.

The BuiltinVariable version is used as a pattern from `call_self_handler()` for `BuiltinVariable(hasattr)`. I think the other version is just used for internal `hasattr(obj, name)` so I renamed that one to `call_obj_hasattr`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593080](https://our.internmc.facebook.com/intern/diff/D68593080)

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1572,7 +1572,7 @@ def call_hasattr(self, tx: "InstructionTranslator", obj, attr):
                 return variables.ConstantVariable(hasattr(obj.fn, name))
-            return obj.call_hasattr(tx, name)
+            return obj.call_obj_hasattr(tx, name)
 
```
Should we have `unimplemented` here? Replacing `call_hasattr` with `call_obj_hasattr` seems somewhat odd and specific to `hasattr`. If you show me the issue, I can try to debug.

anijain2305 reviewed (changes requested):
Seems little odd, so asking for more info.

aorenste committed:
Update on "Remove incorrect BuiltinVariable.call_hasattr()"

BuiltinVariable.call_hasattr() overrides the base class - but actually behaves differently. The base is `obj.call_hasattr(tx, attr)` but BuiltinVariable's version is `<unused>.call_hasattr(tx, obj, attr)`.

The BuiltinVariable version is used as a pattern from `call_self_handler()` for `BuiltinVariable(hasattr)`. I think the other version is just used for internal `hasattr(obj, name)` so I renamed that one to `call_obj_hasattr`.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

Differential Revision: [D68593080](https://our.internmc.facebook.com/intern/diff/D68593080)

[ghstack-poisoned]

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68593080).

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68593080).

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68593080).

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68788166).

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68788166).

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68788166).

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D68875929).
